### DISCUSSION TITLE
Poppler not recognized under Alpine

### DISCUSSION THREAD

**Discussion by perelin** [libvips/libvips#2042] [Q&A]
**Describe the bug**

**In a nutshell**: I want to use VIPS together with Poppler in an Alpine Docker container. After installing `poppler-glib vips-dev vips-tools` VIPS does not seem to recognize the existence of Poppler.

Talking directly to VIPS via the vips cli I cant see any pdf support in the objects list:

```sh
vips -l | grep pdf
```

**Background**: Working on a Go project and including VIPS via https://github.com/davidbyttow/govips 

There for PDF files I always get the error
```sh
Failed to understand image format size=3030 25 50 44
Unsupported image format
```  

Even though govips says it has a loader for pdf during init:

```sh
Vips started with concurrency=1 cache_max_files=100 cache_max_mem=104857600 cache_max=500
Registered image type loader type=png
Registered image type loader type=svg
Registered image type loader type=tiff
Registered image type loader type=webp
Registered image type loader type=gif
Registered image type loader type=jpeg
Registered image type loader type=magick
Registered image type loader type=pdf
```

But when I test in govips for PDF support I get `false`

```go
vips.IsTypeSupported(vips.ImageTypePDF)
```

Have tested JPEG files and they work fine.

**To Reproduce**
* Create a Dockerfile with this content:
```dockerfile
FROM alpine
RUN apk update && apk --no-cache add poppler-glib vips-dev vips-tools
```
* Build image
```sh
docker build -t vipspopplertest .
```
* Notice that there are no entries for pdf operations in the objects list
```sh
docker run vipspopplertest vips -l | grep pdf
```

**Expected behavior**
There are pdf operations listed after `vips -l | grep pdf`

**Environment**
 - OS: Alpine 3.11
 - Vips: 8.8.4

Any pointers what I could try to get some more debugging information on this?

**Comment by jcupitt** [comment]
Hi @perelin,

You need to build libvips with poppler support -- it's a compile time thing. Install poppler-glib-dev and try rebuilding libvips.

**Comment by jcupitt** [comment]
For example:

https://github.com/jcupitt/docker-builds/blob/master/pyvips-alpine/Dockerfile

Though obvs. you'll need to update that a bit.

**Comment by perelin** [comment]
Hi @jcupitt ,

thanks for the quick reply. Ah! I was hopping to get around the "compile by hand" dance :)

I was not able to find a `poppler-glib-dev` package in Alpine:

```sh
$ apk -q search poppler
poppler-qt5
poppler-dev
poppler-glib
poppler-data
poppler
poppler-doc
poppler-utils
zathura-pdf-poppler
poppler-qt5-dev
```

But I used the packages from your Dockerfile example. My Dockerfile now looks like this:

```dockerfile
FROM alpine:latest

ARG VIPS_VERSION=8.10.0
ARG VIPS_URL=https://github.com/libvips/libvips/releases/download

RUN apk update && apk update

RUN apk --no-cache add \
	poppler-glib \
	build-base \
	autoconf \
	automake \
	libtool \
	pkgconf \
	glib-dev \
	expat-dev \
	bc

RUN wget -O- ${VIPS_URL}/v${VIPS_VERSION}/vips-${VIPS_VERSION}.tar.gz | tar xzC /tmp
RUN cd /tmp/vips-${VIPS_VERSION} \
	&& ./configure --prefix=/usr --disable-static --disable-debug \
	&& make V=0 \
	&& make install
```

But still I don´t see any pdf support when running `vips -l | grep pdf`.

I tried substituting `poppler-glib` with `poppler-dev` - same result.

Any hint what to try next?

**Comment by jcupitt** [comment]
You need to check the output of vips configure and add packages until you get "yes" next to pdf in the summary at the end.

I tried my dockerfile and it seems to work fine for me -- I get `vips pdfload`.

**Comment by perelin** [comment]
Yes, that did the trick! Thanks for the support. Any way to contribute?

**Comment by perelin** [comment]
Ah, btw: I use a builder container to compile all that stuff. What files / settings do I need to transfer to the production container to have it work there?

**Comment by jcupitt** [comment]
I would just copy `/usr/local` over. You can trim it a bit perhaps, but it wouldn't shrink much.

There's a "contribute" button at the top of the page, of you'd like to help the project financially.

**Comment by jalevin** [comment]
Hey @jcupitt thanks for your help on this. I was running into the same thing. Looking at my container, I'm at 500mb.. obviously I want to shrink that. 
/usr/local is pretty much empty (I'm not including anything for python)

/usr/lib is 200mb and has the shared libraries. Is that in the ballpark?

**Comment by jcupitt** [comment]
I tried:

```
$ cd GIT/libvips
$ CFLAGS="-g -Wall" CXXFLAGS="-g -Wall" ./autogen.sh --prefix=/home/john/vips
$ make install
$ rm -rf ~/vips
$ du -sk ~/vips
46200	/home/john/vips
```

So 46mb for a debug build.

**Comment by jcupitt** [comment]
I tried a production build, then stripped the libraries and binaries and removed the headers.

```
$ du -sk ~/vips
12520	/home/john/vips
```

**Comment by jalevin** [comment]
Ahh that's awesome. As I was working on the build, I realized, I'm linking against the libs so I can't use a standalone binary. Previous container size before going down this path was 90mb. After getting everything to work correctly and adding pack the shared libs I'm at 199mb. So roughly 110mb of libs. certainly more reasonable to deal with.

Does that seem about right for the shared libs?

Here's the final build step.

```
FROM alpine:latest

RUN apk update && apk add geos glib-dev

RUN apk add libjpeg-turbo \
            poppler-glib \
            libpng \
            libwebp \
            orc \
            tiff \
            libgsf \
            libexif \
            fftw \
            giflib \
            zlib

COPY --from=vipsbuilder /usr/lib/libvip* /usr/lib/
COPY --from=builder /dist/main /
ENTRYPOINT ["/main"]
```
